### PR TITLE
Clean up how `interface` is handled in `QNode` and `qml.execute`

### PR DESCRIPTION
**Pull request by astralcai**
Regarding `numpy` and `autograd`:
- When the parameters are of the `numpy` interface, internally treat it as `interface=None`.
- Does not change the behaviour of treating user specified `interface="numpy"` as using autograd.

Regarding interfaces in general:
- The set of canonical interface names in `INTERFACE_MAP` is expanded to include more specific names such as `jax-jit`, and `tf-autograph`. `_convert_to_interfaces` in `qnode.py` uses a separate `interface_conversion_map` to further map the specific interfaces to their corresponding general interface names that can be passed to the `like` argument of `qml.math.asarray` (e.g. "tf" to "tensorflow", "jax-jit" to "jax").
- In `QNode` and `qml.execute`, every time we get an interface from user input or `qml.math.get_interface`, we map it to a canonical interface name using `INTERFACE_MAP`. Aside from these two scenarios, we assume that the interface name is one of the canonical interface names everywhere else. `QNode.interface` is now assumed to be one of the canonical interface names.
- User input of `interface=None` gets mapped to `numpy` immediately. Internally, `QNode.interface` will never be `None`. It'll be `numpy` for having no interface.
- If `qml.math.get_interface` returns `numpy`, we do not map it to anything. We keep `numpy`.

Collateral bug fix included as well:
- Fixes a bug where a circuit of the `autograd` interfaces sometimes returns results that are not `autograd`.
- Adds `compute_sparse_matrix` to `Hermitian`

[sc-73144]

### CODE REVIEW

File: pennylane/workflow/execution.py [hunk 1]
```diff
@@ -78,7 +78,7 @@
 ]
 
 _mapping_output = (
-    "Numpy",
+    "numpy",
```
Comment (albi3ro) [inline pennylane/workflow/execution.py:81]: Potentially we need to keep this `None`.

Comment (astralcai) [reply pennylane/workflow/execution.py:81]: I thought in `_mapping_output`, `numpy` means `numpy`? What's the difference between `numpy` and `None` here?

Comment (albi3ro) [reply pennylane/workflow/execution.py:81]: The issue if that the interface gets mapped several times. So we end up with `INTERFACE_MAP[INTERFACE_MAP[None]] == "autograd"`.

Comment (astralcai) [reply pennylane/workflow/execution.py:81]: That's a good point. However, do we have handling for the actual `interface=None`?

Comment (astralcai) [reply pennylane/workflow/execution.py:81]: Considering that in `_mapping_output`, `numpy` was previous called `Numpy`, and we never see a key error raised anywhere, I think it's safe to say that the mapped output is never re-mapped.

File: pennylane/devices/legacy_facade.py [hunk 2]
```diff
@@ -25,6 +25,7 @@
 import pennylane as qml
 from pennylane.measurements import MidMeasureMP, Shots
 from pennylane.transforms.core.transform_program import TransformProgram
+from pennylane.workflow.execution import INTERFACE_MAP
```
Comment (albi3ro) [inline pennylane/devices/legacy_facade.py:28]: isn't causing circular imports?

File: pennylane/ops/qubit/observables.py [hunk 3]
```diff
@@ -137,6 +137,10 @@ def compute_matrix(A: TensorLike) -> TensorLike:  # pylint: disable=arguments-di
         Hermitian._validate_input(A)
         return A
 
+    @staticmethod
+    def compute_sparse_matrix(A) -> csr_matrix:  # pylint: disable=arguments-differ
```
Comment (albi3ro) [inline pennylane/ops/qubit/observables.py:141]: ? Why is this needed?

File: pennylane/devices/legacy_facade.py (hunk 2 above)
Comment (astralcai) [reply pennylane/devices/legacy_facade.py:28]: Doesn't look like it. Everything passes.

File: pennylane/ops/qubit/observables.py (hunk 3 above)
Comment (astralcai) [reply pennylane/ops/qubit/observables.py:141]: Because when the interface is `numpy`, [this line](https://github.com/PennyLaneAI/pennylane/blob/5f4169a297605ab6c785eef11828aecb749067d5/pennylane/devices/qubit/measure.py#L199) returns `csr_dot_products`, which uses the sparse matrix.

File: pennylane/workflow/qnode.py [hunk 4]
```diff
@@ -896,12 +906,18 @@ def _execution_component(self, args: tuple, kwargs: dict) -> qml.typing.Result:
 
         execute_kwargs["mcm_config"] = mcm_config
 
+        # Mapping numpy to None here because `qml.execute` will map None back into
+        # numpy. If we do not do this, numpy will become autograd in `qml.execute`.
+        # If the user specified interface="numpy", it would've already been converted to
+        # "autograd", and it wouldn't be affected.
+        interface = self.interface if self.interface != "numpy" else None
```
Comment (albi3ro) [inline pennylane/workflow/qnode.py:913]:
```suggestion
        interface = None if else.interface == "numpy" else self.interface
```

File: pennylane/ops/qubit/observables.py (hunk 3 above)
Comment (albi3ro) [reply pennylane/ops/qubit/observables.py:141]: I might actually call this a bug in DQ.  We should only be using `csr_dot_products` if the observable actually provides a sparse matrix.

I'm wondering now if the performance will shift when we actually start using the optimized numpy logic instead of autograd logic for the simualtions.

Comment (albi3ro) [reply pennylane/ops/qubit/observables.py:141]: But this is potentially a sufficient fix for now.

File: doc/releases/changelog-dev.md [hunk 5]
```diff
@@ -7,18 +7,14 @@
...
 
 * `qml.qchem.excitations` now optionally returns fermionic operators.
-   [(#6171)](https://github.com/PennyLaneAI/pennylane/pull/6171)
+  [(#6171)](https://github.com/PennyLaneAI/pennylane/pull/6171)
 
 * The `diagonalize_measurements` transform now uses a more efficient method of diagonalization 
   when possible, based on the `pauli_rep` of the relevant observables.
   [#6113](https://github.com/PennyLaneAI/pennylane/pull/6113/)
 
-<h4>Capturing and representing hybrid programs</h4>
```
Comment (astralcai) [inline doc/releases/changelog-dev.md:16]: This is a duplicate of the entry directly below.

File: pennylane/workflow/execution.py [hunk 6]
```diff
@@ -135,41 +131,41 @@ def _get_ml_boundary_execute(
         pennylane.QuantumFunctionError if the required package is not installed.
 
     """
-    mapped_interface = INTERFACE_MAP[interface]
```
Comment (astralcai) [inline pennylane/workflow/execution.py:138]: At this point we're assuming that the `interface` is already mapped to one of the canonical interface names.

File: pennylane/workflow/execution.py [hunk 7]
```diff
@@ -479,12 +490,12 @@ def cost_fn(params, x):
         cache = None
 
     # changing this set of conditions causes a bunch of tests to break.
-    no_interface_boundary_required = interface is None or config.gradient_method in {
+    no_interface_boundary_required = interface == "numpy" or config.gradient_method in {
```
Comment (astralcai) [inline pennylane/workflow/execution.py:493]: User-specified `interface=None` had already been mapped to `numpy`.

File: tests/test_qnode.py
Comment (albi3ro) [inline tests/test_qnode.py:1]: Can we add a test here that if the auto-detected interface is numpy, we get numpy/ built-ins out?

Comment (astralcai) [reply tests/test_qnode.py:1]: Will do

Comment (astralcai) [reply tests/test_qnode.py:1]: Done

File: pennylane/devices/legacy_facade.py [hunk 8]
```diff
@@ -367,7 +368,7 @@ def _create_temp_device(self, batch):
             )
             # we already warned about backprop device switching
             new_device = qml.device(
-                backprop_devices[mapped_interface],
+                backprop_devices[interface],
```
Comment (albi3ro) [inline pennylane/devices/legacy_facade.py:371]: will potentially have more merge conflicts soon 😢 . sorry about that.

File: pennylane/ops/qubit/observables.py (hunk 3 above)
Comment (albi3ro) [reply pennylane/ops/qubit/observables.py:141]: #6264

File: pennylane/workflow/execution.py (hunk 6 above)
Comment (albi3ro) [reply pennylane/workflow/execution.py:138]: 👍

**Review by albi3ro — APPROVED** [review]
🎉  Very happy to finally be getting this done 🚀 🎉

File: doc/releases/changelog-dev.md [hunk 9]
```diff
@@ -10,18 +10,14 @@
    [(#6061)](https://github.com/PennyLaneAI/pennylane/pull/6061)
 
 * `qml.qchem.excitations` now optionally returns fermionic operators.
-   [(#6171)](https://github.com/PennyLaneAI/pennylane/pull/6171)
+  [(#6171)](https://github.com/PennyLaneAI/pennylane/pull/6171)
```
Comment (PietropaoloFrisoni) [inline doc/releases/changelog-dev.md:13]: Potentially the entry above needs the same correction XD

File: pennylane/workflow/execution.py [hunk 10]
```diff
@@ -78,30 +75,29 @@
...
     "jax",
     "jax",
     "torch",
     "torch",
     "tf",
     "tf",
-    "tf",
-    "tf",
+    "tf-autograph",
+    "tf-autograph",
```
Comment (PietropaoloFrisoni) [inline pennylane/workflow/execution.py:92]: It's probably a trivial question, but why are some entries repeated twice?

**Review by PietropaoloFrisoni — APPROVED** [review]
LGTM! Just a curiosity and a non-blocking observation
